lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,206 @@
+-- +------------+-------------+------------------------------------------------+
+-- | p_apparent | @ [0, 1] @  | apparent performance                           |
+-- +------------+-------------+------------------------------------------------+
+-- | s          | @ [0, ∞) @  | relative stake of the pool leader (i.e pledge) |
+-- +------------+-------------+------------------------------------------------+
+-- | σ          | @ [0, 1] @  | total relative stake of the pool               |
+-- +------------+-------------+------------------------------------------------+
+module Cardano.Pool.Ranking
+    (
+      -- * Formulas
+      desirability
+
+    , saturatedPoolRewards
+    , saturatedPoolSize
+
+      -- * Types
+    , EpochConstants (..)
+    , Pool (..)
+    , RelativeStakeOf (..)
+    , mkRelativeStake
+    , Lovelace (..)
+    , Margin
+    , unsafeMkMargin
+    , getMargin
+    , NonNegative (..)
+    , Positive (..)
+    , unsafeToPositive
+    , unsafeToNonNegative
+    )
+    where
+
+import Prelude
+
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( Symbol )
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+-- | Non-myopic pool desirability according to section 5.6.1.
+--
+-- Is /not/ affected by oversaturation nor pool stake in general.
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = getMargin $ margin pool
+    c = getNonNegative $ getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getNonNegative $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , optimalNumberOfPools :: Positive Int
+      -- ^ k
```

Comment:
`k` is referred to as the _"desired number of pools"_, which is different from an _"optimal"_ number. Rather, we chose to fix `k` and then, determine what other parameters should be. But, chosing `k=10` or `k=100` or `k=1000` is rather arbitrary and simply witnesses of how decentralized we want the network to be.